### PR TITLE
Refactor Cleanroom and CharcoalPileIgnitor Structure Check

### LINKED ISSUE #2419: Allow a way for modpacks to add additional blocks that can be part of cleanroom floor/walls
### Cross-mod Integration

e.g. OpenBlocks Elevator, Tranvel Anchors

### Feature Description

It you could include things like elevators or travel anchors (perhaps up to a small (2-4) limit) could be included in the floor or walls of a cleanroom. One way to handle this would be to have a block tag of extra blocks that can be included in a cleanroom structure. The tag could either be left blank by default, or populated with elevators/travel anchors from some common mods.

### PR DESCRIPTION
**Pull request by YoungOnionMC**
## What
Refactors the entire structure builder for both cleanroom and charcoal pile ignitor multis
Adds in a map for valid blocks in the cleanroom floor
Fixes #2419 

## Implementation Details
The layer StringBuilders are done from bottom up instead of left to right like it used to

## Outcome
More flexibility in building a cleanroom

### CODE REVIEW

File: src/main/java/com/gregtechceu/gtceu/config/ConfigHolder.java [hunk 1]
```diff
@@ -441,6 +441,11 @@ public static class MachineConfigs {
                 "This does nothing if enableCleanroom is false.", "Default: false" })
         public boolean cleanMultiblocks = false;
         @Configurable
+        @Configurable.Comment({ "List of BlockStates that are valid as part of the cleanroom",
+                "Entries must be in a fully qualified format. For example: ",
+                "Default: none" })
+        public String[] cleanroomFloorBlocks = new String[0];
+        @Configurable
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/config/ConfigHolder.java]: make this a tag instead of a config

**Review by screret — CHANGES_REQUESTED** [review]

File: src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/CleanroomMachine.java [hunk 2]
```diff
@@ -291,7 +286,7 @@ public boolean isBlockFloor(@NotNull Level world, @NotNull BlockPos.MutableBlock
         var state = world.getBlockState(pos.move(direction));
         return state == GTBlocks.PLASTCRETE.getDefaultState() ||
                 state == GTBlocks.CLEANROOM_GLASS.getDefaultState() ||
-                validFloorStates.contains(state);
+                state.is(CustomTags.CLEANROOM_FLOORS);
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/CleanroomMachine.java]: remove the 1st 2 cases, useless

File: src/main/java/com/gregtechceu/gtceu/data/tags/BlockTagLoader.java [hunk 3]
```diff
@@ -43,6 +44,8 @@ public static void init(RegistrateTagsProvider<Block> provider) {
         provider.addTag(BlockTags.MINEABLE_WITH_AXE)
                 .add(TagEntry.element(GTMachines.WOODEN_DRUM.getId()))
                 .add(TagEntry.element(GTMachines.WOODEN_CRATE.getId()));
+
+        create(provider, CustomTags.CLEANROOM_FLOORS, GTBlocks.PLASTCRETE.get(), GTBlocks.CLEANROOM_GLASS.get());
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/data/tags/BlockTagLoader.java:48]: cleanroom glass shouldn't be a valid floor block, only a wall block

**Review by screret — CHANGES_REQUESTED** [review]

File: src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/CleanroomMachine.java [hunk 4]
```diff
@@ -418,6 +407,14 @@ protected static TraceabilityPredicate doorPredicate() {
...
     }
 
+    private TraceabilityPredicate getValidFloorBlocks() {
+        return Predicates.custom(blockWorldState -> blockWorldState.getBlockState().is(CustomTags.CLEANROOM_FLOORS),
+                () -> BuiltInRegistries.BLOCK.getTag(CustomTags.CLEANROOM_FLOORS)
+                        .get().stream()
+                        .map(h -> new BlockInfo(h.get()))
+                        .toArray(BlockInfo[]::new));
+    }
+
```
Comment (screret) [inline src/main/java/com/gregtechceu/gtceu/common/machine/multiblock/electric/CleanroomMachine.java]: actually this could just use `Predicates.blockTag`

**Review by screret — APPROVED** [review]

**Review by screret — APPROVED** [review]

File: src/generated/resources/data/gtceu/tags/blocks/cleanroom_floors.json [hunk 5]
```diff
@@ -0,0 +1,6 @@
+{
+  "values": [
+    "gtceu:plascrete",
+    "gtceu:cleanroom_glass"
```
Comment (tomprince) [inline src/generated/resources/data/gtceu/tags/blocks/cleanroom_floors.json:4]: I was testing this on 1.20.1-1.5.5 SNAPSHOT 20241206 and I'm not sure that these blocks should be here.

After adding a couple of more block to the tag, I tried placing two of them in the foor, one in the center, and one beside it. With only the center block replaced, the structure forms, but if I replace one of the surronding blocks it doesn't. However, if I remove these blocks from the tag, I can replace the surronding blocks and still have the structure form.

I suspect what is going on is that the blocks above are eating the global max, before falling back to the other alternatives.

**Pull request merged**